### PR TITLE
Route Builder Pattern

### LINKED ISSUE #52: Route Builder Pattern
## Is your feature request related to a problem? Please describe

Route have a lot of configuration and could be a good candidate for the builder pattern.

## Describe the solution

We could support creating a route with following options:

- Raw using std handler
- Using patron handler
- Using trace
- Using auth
- etc.

```go
New("GET", "/user").WithStdHandler(hnd).Create()
```
or
```go
New("GET", "/user").WithHandler(hnd).WithAuth(auth).WithTrace(cfg).Create()
```

### LINKED ISSUE #97: Add multiple []Route slices and assure non-duplication
## Is your feature request related to a problem? Please describe

One of the ways to design routes in your app is to declare URL mappings in routes themselves, and then pass all of them to Patron.

So it looks something like this (pseudocode):
```
catHTTP := NewCatHTTP()
dogHTTP := NewDogHTTP()

var catRoutes, dogRoutes []http.Route
catRoutes = catHTTP.Routes()
dogRoutes = catHTTP.Routes()

patron.New( ... patron.Routes(
   combineRoutes(catRoutes, dogRoutes)
) ... )

...

func combineRoutes(rrs ...[]http.Route) ([]http.Route, error) { ... }
```

The proposal is to implement this method in Patron.

## Describe the solution

https://github.com/taxibeat/pricing-surge-service/pull/12/files#diff-d6344a9314798b136af53b3c99e586b9R60

### PR DESCRIPTION
**Pull request by mantzas**
## Which problem is this PR solving?

Closes #52.
Closes #97.

## Short description of the changes

- [x] Added a route builder
- [x] Used routebuilder in HTTP component
- [x] Check for route duplicates
- [x] Use routes builder in service

### CODE REVIEW

File: sync/http/route.go [hunk 1]
```diff
@@ -1,130 +1,197 @@
 package http
```
Comment (drakos74) [inline sync/http/route.go:2]: would it make sense to remove the 'Route' part from the method names ? It s part of the package.

File: sync/http/route.go [hunk 2]
```diff
@@ -1,130 +1,197 @@
...
-	return NewRoute(p, http.MethodOptions, pr, trace, nil, mm...)
+// WithMethodPost HTTP method.
+func (rb *RouteBuilder) WithMethodPost() *RouteBuilder {
+	rb.method = http.MethodPost
+	return rb
 }
 
-// NewRoute creates a new route from a generic handler with auth capability.
-func NewRoute(p string, m string, pr sync.ProcessorFunc, trace bool, auth auth.Authenticator, mm ...MiddlewareFunc) Route {
+// WithMethodPut HTTP method.
```
Comment (drakos74) [inline sync/http/route.go]: Given that now we can call these methods sequentially, but we can only have one method, would it make sense to check if rb.method is set , and potentially throw an error ? We might also consider changing the name of these methods to remove the "WithMethod" part. What do you think ?

**Comment by drakos74** [comment]
would this also cover this issue ? https://github.com/beatlabs/patron/issues/97

File: sync/http/route.go (hunk 1 above)
Comment (mantzas) [reply sync/http/route.go:2]: the package is name http. What are you referring to?

File: sync/http/route.go (hunk 2 above)
Comment (mantzas) [reply sync/http/route.go]: About the error i agree!
What is your proposal for the naming?

**Comment by mantzas** [comment]
> would this also cover this issue ? #97

Yes we could! On it

File: sync/http/route.go (hunk 2 above)
Comment (drakos74) [reply sync/http/route.go]: maybe just 'Put' , 'Get' etc ... ?

Comment (mantzas) [reply sync/http/route.go]: I did rename from `WithMethodGet()` to `MethodGet()`. Using just `Get()` looked like a getter method and was not clear as an API.

**Review by drakos74 — APPROVED** [review]

**Review by sotirispl — APPROVED** [review]

**Pull request merged**